test/integration/replicaset/replicaset_test.go

```diff
@@ -272,246 +280,595 @@ func TestAdoption(t *testing.T) {
 	defer framework.DeleteTestingNamespace(ns, s, t)
+	labelMap := map[string]string{"foo": "bar"}
 	rs := newRS("rs", ns.Name, 2)
-	pod1 := newMatchingPod("pod1", ns.Name)
-	pod1.Labels["uniqueKey"] = "1"
-	pod2 := newMatchingPod("pod2", ns.Name)
-	pod2.Labels["uniqueKey"] = "2"
-	createRSsPods(t, clientSet, []*v1beta1.ReplicaSet{rs}, []*v1.Pod{pod1, pod2}, ns.Name)
+	rs.Spec.Selector.MatchLabels = labelMap
+	rs.Spec.Template.Labels = labelMap
+	rss, _ := createRSsPods(t, c, []*v1beta1.ReplicaSet{rs}, []*v1.Pod{}, ns.Name)
+	rs = rss[0]
 
 	stopCh := make(chan struct{})
 	informers.Start(stopCh)
-	waitToObservePods(t, podInformer, 2)
 	go rm.Run(5, stopCh)
-	waitRSStable(t, clientSet, rs, ns.Name)
-
-	// change the rs's selector to match both pods
-	patch := `{"spec":{"selector":{"matchLabels": {"uniqueKey":"1"}},"template":{"metadata":{"labels":{"uniqueKey":"1"}}}}}`
-	rsClient := clientSet.Extensions().ReplicaSets(ns.Name)
-	rs, err := rsClient.Patch(rs.Name, types.StrategicMergePatchType, []byte(patch))
+	defer close(stopCh)
+	waitRSStable(t, c, rs, ns.Name)
+
+	testStatusReplicas(t, c, rs, ns.Name, 3)
+	testStatusReplicas(t, c, rs, ns.Name, 0)
+	testStatusReplicas(t, c, rs, ns.Name, 2)
+
+	// Update 2 pods that RS created such that one is failed and one is being deleted
+	// Ensure both are being created again in the end
+	podClient := c.Core().Pods(ns.Name)
+	podSelector := labels.Set{"foo": "bar"}.AsSelector()
+	options := metav1.ListOptions{LabelSelector: podSelector.String()}
+	pods, err := podClient.List(options)
 	if err != nil {
-		t.Fatalf("Failed to patch replica set: %v", err)
+		t.Fatalf("Failed obtaining a list of pods that match the pod labels %v: %v", labelMap, err)
 	}
-	t.Logf("patched rs = %#v", rs)
-	// wait for the rs to create one more pod
-	if err := wait.Poll(10*time.Second, 60*time.Second, func() (bool, error) {
-		return verifyRemainingObjects(t, clientSet, ns.Name, 1, 3)
+	if pods == nil {
+		t.Fatalf("Obtained a nil list of pods")
+	}
+	if len(pods.Items) != 2 {
+		t.Fatalf("len(pods) = %v, want %v", len(pods.Items), 2)
+	}
+
+	failedPod := &pods.Items[0]
+	deletingPod := &pods.Items[1]
+	if err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		newPod, err := podClient.Get(failedPod.Name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		newPod.Status.Phase = v1.PodFailed
+		_, err = podClient.Update(newPod)
+		return err
 	}); err != nil {
-		t.Fatal(err)
+		t.Fatalf("Failed to set .Status.Phase of pod %s: %v", failedPod.Name, err)
 	}
-	podClient := clientSet.Core().Pods(ns.Name)
-	pod2, err = podClient.Get(pod2.Name, metav1.GetOptions{})
-	if err != nil {
-		t.Fatalf("Failed to get pod2: %v", err)
+	if err = retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		newPod, err := podClient.Get(deletingPod.Name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+		return c.Core().Pods(ns.Name).Delete(newPod.Name, &metav1.DeleteOptions{})
```

Comment:
Interesting - thanks for the clarification! I don't think I am able to find this logic error myself.

Reply:
I realized that I initially did try to set a `deletionTimestamp`, but the field is meant to be changed via deletion only (the field is immutable except during deletion). How should we simulate the recovery without deleting the pod?

The validation error: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/api/validation/objectmeta.go#L287

Reply:
Oops! I was still thinking in terms of unit tests where there is no validation. :)

Try keeping it as a Delete() call, but set GracePeriodSeconds in DeleteOptions to something very large.

I think we need to be careful though. We don't want to hang the test cleanup step while it waits for the Pod to actually get deleted. You might need to add a step after you check the RS behavior, to delete the Pod again with GracePeriodSeconds=0.

Reply:
Sound good to me!
I will have 2 `Delete()` calls:
(1) first with very large `GracePeriodSeconds` to simulate change to `DeletionTimestamp`
(2) second with `GracePeriodSeconds` = 0 to delete the pod immediately

Reply:
@enisoc I set both `GracePeriodSeconds` of `DeleteOptions` and `DeletionGracePeriodSeconds` to very large values.
```
gracePeriodSeconds := int64(3600)
pod.DeletionGracePeriodSeconds = &gracePeriodSeconds
return c.Core().Pods(ns.Name).Delete(pod.Name, &metav1.DeleteOptions{
 		GracePeriodSeconds: &gracePeriodSeconds,
})
```
Afterwards, I tried to print out `DeletionTimestamp` of the pod, but it is `nil`. It looks like `DeletionTimestamp` is not allowed to be set explicitly in the integration test environment.

Reply:
@enisoc Note that `DeletionGracePeriodSeconds` is also an immutable field for validation purpose. The code snippet above does not return any validation error because the pod is not fetched again after the line `pod.DeletionGracePeriodSeconds = &gracePeriodSeconds`. 

Based on my observation, regardless what `GracePeriodSeconds` is, the `Delete()` call will delete the pod _**immediately**_ (or **_hide_** the pod from being accessed via calls such as `GET` and `LIST`). In fact, this should not be a blocking issue - the test wants to verify a new pod is created to replace the deleting pod, not whether a pod can still be fetched / listed within `GracePeriodSeconds` period. IMO, the latter should be covered by a unit test that is not restricted by validation rules as it relies on `DeletionTimestamp`. 

It seems to me that without being able to have a non-nil `DeletionTimestamp` will _**alter**_ the original unit test's objective. That's the reason I asked about it although it is, technically speaking, does not affect the test to pass (in fact, i was able to get the test passed a long time ago).